src/main/java/reactor/rabbitmq/ExceptionHandlers.java

```diff
@@ -0,0 +1,155 @@
+import java.util.concurrent.Callable;
+import java.util.function.BiConsumer;
+import java.util.function.Predicate;
+
+/**
+ *
+ */
+public class ExceptionHandlers {
+
+    public static Predicate<Throwable> CONNECTION_RECOVERY_PREDICATE = new ConnectionRecoveryTriggeringPredicate();
+
+    public static class ConnectionRecoveryTriggeringPredicate implements Predicate<Throwable> {
+
+        @Override
+        public boolean test(Throwable throwable) {
+            if (throwable instanceof ShutdownSignalException) {
+                ShutdownSignalException sse = (ShutdownSignalException) throwable;
+                return !sse.isInitiatedByApplication() || (sse.getCause() instanceof MissedHeartbeatException);
+            }
+            return false;
+        }
+    }
+
+    public static class ExceptionPredicate implements Predicate<Throwable> {
+
+        private final Map<Class<? extends Throwable>, Boolean> retryableExceptions;
+
+        public ExceptionPredicate(Map<Class<? extends Throwable>, Boolean> retryableExceptions) {
+            this.retryableExceptions = retryableExceptions;
+        }
+
+        @Override
+        public boolean test(Throwable throwable) {
+            for (Map.Entry<Class<? extends Throwable>, Boolean> retryableException : retryableExceptions.entrySet()) {
+                if (retryableException.getKey().isAssignableFrom(throwable.getClass()) &&
+                    retryableException.getValue().booleanValue()) {
+                    return true;
+                }
+            }
+            return false;
+        }
+    }
+
+    public static class SimpleRetryTemplate {
+
+        private final long timeout;
+        private final long waitingTime;
+
+        private final Predicate<Throwable> predicate;
+
+        public SimpleRetryTemplate(Duration timeout, Duration waitingTime, Predicate<Throwable> predicate) {
+            if (timeout == null || timeout.isNegative() || timeout.isZero()) {
+                throw new IllegalArgumentException("Timeout must be greater than 0");
+            }
+            if (waitingTime == null || timeout.isNegative() || timeout.isNegative()) {
+                throw new IllegalArgumentException("Waiting time must be greater than 0");
+            }
+            if (timeout.compareTo(waitingTime) <= 0) {
+                throw new IllegalArgumentException("Timeout must be greater than waiting time");
+            }
+            if (predicate == null) {
+                throw new NullPointerException("Predicate cannot be null");
+            }
+            this.timeout = timeout.toMillis();
+            this.waitingTime = waitingTime.toMillis();
+            this.predicate = predicate;
+        }
+
+        public void retry(Callable<Void> operation, Exception e) {
+            if (predicate.test(e)) {
+                int elapsedTime = 0;
+                while (elapsedTime < timeout) {
+                    try {
+                        Thread.sleep(waitingTime);
```

Comment:
This is the only element that I have a doubt about. Sleeping the `Thread` could make sense if you're 100% sure to be in a rabbit-managed thread at this point, but you could also schedule the `operation` on the appropriate `Scheduler`?

Also, have you looked at the baked in `retry` operators in Reactor? In 3.2 we're introducing a "sane default" exponential backoff retry, that is already present in `reactor-extra` in the 3.1.x line.

Reply:
Exception handlers are typically called from Reactor-management threads:
https://github.com/reactor/reactor-rabbitmq/blob/be0e0a80829603e0d342303e1c5dfd5ab96b86bf/src/main/java/reactor/rabbitmq/Sender.java#L115
https://github.com/reactor/reactor-rabbitmq/blob/master/src/main/java/reactor/rabbitmq/Sender.java#L494
https://github.com/reactor/reactor-rabbitmq/blob/master/src/main/java/reactor/rabbitmq/Receiver.java#L129

The point here is to handle network failures in the most appropriate way for a default. If the connection is dead when sending something to the broker (message or ack), all subsequent write operations are likely to fail too until the connection automatically recovers, hence the synchronous retry.

Scheduling the retry may work too, but it'll also make the exception handling harder to reason about. WDYT @simonbasle?

Built-in Reactor retry may be useful too, I'll have a look at it.